File: 3677.count-binary-palindromic-numbers.py
#
# @lc app=leetcode id=3677 lang=python3
#
# [3677] Count Binary Palindromic Numbers
#

# @lc code=start
class Solution:
    def countBinaryPalindromes(self, n: int) -> int:
        if n == 0:
            return 1
        length = len(bin(n)[2:])
        count = 1  # for 0
        for l in range(1, length):
            count += 1 << ((l - 1) // 2)
        half_len = (length + 1) // 2
        low = 1 << (half_len - 1)
        high = (1 << half_len) - 1
        def gen_pal(left_val: int) -> int:
            full = 0
            p_l = 1 << (length - 1)
            p_r = 1
            h = half_len
            for i in range(h):
                bit = (left_val >> (h - 1 - i)) & 1
                full += bit * p_l
                if i < length - 1 - i:
                    full += bit * p_r
                p_l >>= 1
                p_r <<= 1
            return full
        l, r = low, high
        while l <= r:
            mid = l + (r - l) // 2
            if gen_pal(mid) <= n:
                l = mid + 1
            else:
                r = mid - 1
        add = r - low + 1 if r >= low else 0
        count += add
        return count

# @lc code=end

File: 3677.count-binary-palindromic-numbers.java
/*
 * @lc app=leetcode id=3677 lang=java
 *
 * [3677] Count Binary Palindromic Numbers
 */
// @lc code=start
class Solution {
    public int countBinaryPalindromes(long n) {
        int count = (n >= 0) ? 1 : 0; // count 0 if in range
        int maxBits = Long.toBinaryString(n).length();
        for (int len = 1; len <= maxBits; len++) {
            int half = (len + 1) / 2;
            long start = 1L << (half - 1);
            long end = 1L << half;
            for (long i = start; i < end; i++) {
                long pal = createPalindrome(i, len % 2 == 1);
                if (pal > n) break;
                if (pal != 0) count++;
            }
        }
        return count;
    }
    private long createPalindrome(long x, boolean odd) {
        long res = x;
        if (odd) x >>= 1;
        while (x > 0) {
            res = (res << 1) | (x & 1);
            x >>= 1;
        }
        return res;
    }
}
// @lc code=end